# Dummy Variables
In this topic we would be looking at how to predict priices of properties given the fact we have the location of those properties in 'string' formats. To achieve this we would need to assign dummy variables to the string formats in the town columns.

In [1]:
import pandas as pd
df = pd.read_csv(r"C:\Users\Anotijakes\Documents\DATAS\CSV DATA\state_prices.csv")
df

town  area   price
0  abeokuta  2600  550000
1  abeokuta  3000  565000
2  abeokuta  4000  725000
3       jos  2600  585000
4       jos  2800  615000
5       jos  3200  630000
6     Abuja  2600  575000
7     Abuja  3300  650000
8     Abuja  3600  710000

There is a format  used in assigning dummy varriables and it's know as 'one hot encoding'. 
where binary numbers are used to  replace  the towns in the columns
the function used is 'get_dummies(column name)'.

In [2]:
dummies = pd.get_dummies(df.town)
dummies

Abuja  abeokuta  jos
0      0         1    0
1      0         1    0
2      0         1    0
3      0         0    1
4      0         0    1
5      0         0    1
6      1         0    0
7      1         0    0
8      1         0    0

In [3]:
#next we add the dummie table created to the original data frame using the concatenate function.
#so we would be joing the df data frame and the 'dummies' data frame and specify if it's column or rows
merged = pd.concat([df,dummies], axis = 'columns')
merged

town  area   price  Abuja  abeokuta  jos
0  abeokuta  2600  550000      0         1    0
1  abeokuta  3000  565000      0         1    0
2  abeokuta  4000  725000      0         1    0
3       jos  2600  585000      0         0    1
4       jos  2800  615000      0         0    1
5       jos  3200  630000      0         0    1
6     Abuja  2600  575000      1         0    0
7     Abuja  3300  650000      1         0    0
8     Abuja  3600  710000      1         0    0

In [4]:
#next  we would need to drop the town column and one of the states column.
final = merged.drop(['town','Abuja'], axis = 'columns')
final

area   price  abeokuta  jos
0  2600  550000         1    0
1  3000  565000         1    0
2  4000  725000         1    0
3  2600  585000         0    1
4  2800  615000         0    1
5  3200  630000         0    1
6  2600  575000         0    0
7  3300  650000         0    0
8  3600  710000         0    0

next we import our sklearn linear regression model for our predictions

In [5]:
from sklearn.linear_model import LinearRegression
model = LinearRegression()

In other to find our 'x' , our x is all the columns except our price column. Our price column as we all know is our 'y'
so we drop just the price column to get our x.

In [6]:
x = final.drop("price", axis = 'columns')
x

area  abeokuta  jos
0  2600         1    0
1  3000         1    0
2  4000         1    0
3  2600         0    1
4  2800         0    1
5  3200         0    1
6  2600         0    0
7  3300         0    0
8  3600         0    0

In [7]:
y = final.price
y

0    550000
1    565000
2    725000
3    585000
4    615000
5    630000
6    575000
7    650000
8    710000
Name: price, dtype: int64

In [8]:
#now we have our x and y, it's time to make the predictions.
# now let's train our model
model.fit(x,y)

LinearRegression()

In [9]:
# now let's predict the price of 2800 sqt home in abeokuta
# you will need to supply the parameters for your x
# using abeokuta as '1' and jos as '0' in the onehotencoding format.
model.predict([[2800,1,0]])

C:\Users\Anotijakes\anaconda3\lib\site-packages\sklearn\base.py:450: UserWarning: X does not have valid feature names, but LinearRegression was fitted with feature names
  warnings.warn(


array([563485.42458809])

In [10]:
# next let's do 3400 sqt in jos.
model.predict(([[3400,0,1]]))

C:\Users\Anotijakes\anaconda3\lib\site-packages\sklearn\base.py:450: UserWarning: X does not have valid feature names, but LinearRegression was fitted with feature names
  warnings.warn(


array([676463.87832699])

In [11]:
#  in other to predict for abuja you can imput '0' and '0' the the last two arguments,, checking how it is from the
# original table
# Let's check for 2600 sqt
model.predict(([[2600,0,0]]))

C:\Users\Anotijakes\anaconda3\lib\site-packages\sklearn\base.py:450: UserWarning: X does not have valid feature names, but LinearRegression was fitted with feature names
  warnings.warn(


array([574382.12927758])

As we can see, from the original table, the priice was 575,000, but in the prediction it is 574382.1 which is close but not sooo close. in other to check how accurate your predictions are you use the 'score()' function.

In [12]:
model.score(x,y)
# the accuracy of the prediction is 94%. 
# x is the pridicted column, y is the actual column. The both of them are compared.

0.9428182771732004

Now we are going to be using sklearn onehotencoder to do the same thing

In [13]:
# first we do labor encoding on the town column. 
# this is still our original column
df

town  area   price
0  abeokuta  2600  550000
1  abeokuta  3000  565000
2  abeokuta  4000  725000
3       jos  2600  585000
4       jos  2800  615000
5       jos  3200  630000
6     Abuja  2600  575000
7     Abuja  3300  650000
8     Abuja  3600  710000

In [25]:
# we import sklearn label encoder
from sklearn.preprocessing import LabelEncoder
le = LabelEncoder()

In [26]:
# let's connect the variabe 'dfle' to the data frame
dfle = df
le.fit(dfle.town)

LabelEncoder()

In [27]:
# assign it to ur town column
dfle.town = le.fit(dfle.town)

In [28]:
dfle
# unfortunately from the video it shows numbers but here it shows LabelEncoder. 
# so i'll stop hre

town  area   price
0  LabelEncoder()  2600  550000
1  LabelEncoder()  3000  565000
2  LabelEncoder()  4000  725000
3  LabelEncoder()  2600  585000
4  LabelEncoder()  2800  615000
5  LabelEncoder()  3200  630000
6  LabelEncoder()  2600  575000
7  LabelEncoder()  3300  650000
8  LabelEncoder()  3600  710000